arangod/RocksDBEngine/RocksDBGeoIndex.cpp

```diff
@@ -263,6 +420,205 @@ class RDBNearIterator final : public IndexIterator {
+                           if (res.fail() ||
+                               (ft == geo::FilterType::CONTAINS &&
+                                !filter.contains(&test)) ||
+                               (ft == geo::FilterType::INTERSECTS &&
+                                !filter.intersects(&test))) {
+                             result = false;
+                             return false;
+                           }
+                           return true;
+                           // geo index never needs to observe own writes
+                         },
+                         ReadOwnWrites::no)
+                     .ok()) {
+              return false;
+            }
+            if (!result) {
+              return false;
+            }
+          }
+
+          cb(docid);  // return result
+          return true;
+        },
+        limit);
+  }
+
+  void resetImpl() override { _covering.reset(); }
+
+ private:
+  void performScan() {
+    rocksdb::Comparator const* cmp = _index->comparator();
+    // list of sorted intervals to scan
+    if (_gotIntervals) {
+      _scan = _covering.intervals();
+      _gotIntervals = true;
+      _scanningInterval = 0;
+    }
+    while (_scanningInterval < _scan.size()) {
+      geo::Interval const& it = _scan[_scanningInterval];
+      TRI_ASSERT(it.range_min <= it.range_max);
+      RocksDBKeyBounds bds = RocksDBKeyBounds::GeoIndex(
+          _index->objectId(), it.range_min.id(), it.range_max.id());
+
+      // intervals are sorted and likely consecutive, try to avoid seeks
+      // by checking whether we are in the range already
+      bool seek = true;
+      if (_scanningInterval > 0) {
+        TRI_ASSERT(_scan[_scanningInterval - 1].range_max < it.range_min);
+        if (!_iter->Valid()) {  // no more valid keys after this
+          break;
+        } else if (cmp->Compare(_iter->key(), bds.end()) > 0) {
+          continue;  // beyond range already
+        } else if (cmp->Compare(bds.start(), _iter->key()) <= 0) {
+          seek = false;  // already in range: min <= key <= max
+          TRI_ASSERT(cmp->Compare(_iter->key(), bds.end()) <= 0);
+        } else {  // cursor is positioned below min range key
+          TRI_ASSERT(cmp->Compare(_iter->key(), bds.start()) < 0);
+          int k = 10;  // try to catch the range
+          while (k > 0 && _iter->Valid() &&
+                 cmp->Compare(_iter->key(), bds.start()) < 0) {
+            _iter->Next();
+            --k;
+          }
+          seek =
+              !_iter->Valid() || (cmp->Compare(_iter->key(), bds.start()) < 0);
+        }
+      }
+
+      if (seek) {  // try to avoid seeking at all cost
+        _iter->Seek(bds.start());
+      }
+
+      while (_iter->Valid() && cmp->Compare(_iter->key(), bds.end()) <= 0) {
+        _covering.reportFound(RocksDBKey::indexDocumentId(_iter->key()),
+                              RocksDBValue::centroid(_iter->value()));
+        _iter->Next();
+      }
```

Comment:
We limit to 1024 documents being found (and accepted). We could improve this, but it would be a logic change. I suggest we leave this for a later performance optimization PR.